### PR TITLE
Fix openai flavor impelementation to support formatting messages with prediction input

### PR DESCRIPTION
**Pull request by harupy**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Fix `openai` flavor implementation to support formatting `messages` with prediction input.

## How is this patch tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [x] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/openai/utils.py [hunk 1]
```diff
@@ -65,8 +66,9 @@ def request(*args, **kwargs):
         else:
             url = kwargs.get("url")
 
-        if "chat/completions" in url:
-            return _mock_chat_completion_response()
+        if re.match(r"^https://api\.openai\.com/v\d+/chat/completions$", url):
+            messages = json.loads(kwargs.get("data")).get("messages")
+            return _mock_chat_completion_response(content=json.dumps(messages))
```
Comment (harupy) [inline mlflow/openai/utils.py:71]: To check a request is sent with an expected `messages`, return a dumped `messages` as a response.

File: mlflow/openai/api_request_parallel_processor.py [hunk 2]
```diff
@@ -208,8 +208,11 @@ def process_api_requests(
     last_update_time = time.time()
     results: list[tuple[int, OpenAIObject]] = []
     requests_iter = enumerate(requests)
+    last_index = len(requests) - 1
+    requests_exhausted = False
     with ThreadPoolExecutor(max_workers=max_workers) as executor:
         while True:
+            time.sleep(0.001)
```
Comment (harupy) [inline mlflow/openai/api_request_parallel_processor.py:215]: Make a short sleep to avoid busy-waiting

File: mlflow/openai/__init__.py [hunk 3]
```diff
@@ -191,6 +193,37 @@ def _log_secrets_yaml(local_model_dir, scope):
...
+            )
+        )
+    )
+
+
+def _get_input_schema(messages):
+    if messages:
+        variables = _collect_variables(messages)
+        if len(variables) == 1:
+            return Schema([ColSpec(name=None, type="string")])
```
Comment (harupy) [inline mlflow/openai/__init__.py:214]: Is this correct?

File: mlflow/openai/__init__.py [hunk 4]
```diff
@@ -154,7 +156,7 @@ def _get_task_name(task):
         )
 
 
-def get_openai_package_version():
+def _get_openai_package_version():
```
Comment (harupy) [inline mlflow/openai/__init__.py:159]: This should not be a public API.

File: mlflow/openai/__init__.py [hunk 5]
```diff
@@ -437,37 +495,52 @@ def __init__(self, model):
...
+        variables = _collect_variables(messages)
+        if variables:
+            messages_list = self.format_messages(messages, self.get_params_list(data, variables))
+        else:
+            iter_string_columns = (c for c, v in data.iloc[0].items() if isinstance(v, str))
+            first_string_column = next(iter_string_columns, None)
+            if first_string_column is None:
                 raise mlflow.MlflowException.invalid_parameter_value(
-                    "Input list of dictionaries must contain keys 'content' and 'role'",
+                    "No string columns found in the input data"
```
Comment (BenWilson2) [inline mlflow/openai/__init__.py:530]: Could we make this message more clear with an example of the supported input types? "You must supply either <type x> or <type y> in order to use this API for inference"

File: mlflow/openai/__init__.py [hunk 6]
```diff
@@ -429,6 +470,23 @@ def _has_content_and_role(d):
...
+class _Message:
+    def __init__(self, message):
+        self.raw_content = message.get("content")
+        self.role = message.get("role")
+        self.variables = _parse_format_fields(self.raw_content)
+
+    def format(self, **params):
+        if missing_params := set(self.variables) - set(params):
+            raise mlflow.MlflowException.invalid_parameter_value(
+                f"Missing parameters: {missing_params}"
```
Comment (BenWilson2) [inline mlflow/openai/__init__.py:482]: Can this message be a little more verbose to let the users know the structure that is expected?

File: mlflow/openai/__init__.py [hunk 7]
```diff
@@ -437,37 +495,52 @@ def __init__(self, model):
...
 
-        if isinstance(data, pd.DataFrame):
-            if "content" not in data or "role" not in data:
-                raise mlflow.MlflowException.invalid_parameter_value(
-                    "Input dataframe must contain columns 'content' and 'role'",
-                )
-            messages = data.to_dict(orient="records")
-        elif isinstance(data, list) and all(isinstance(d, dict) for d in data):
-            if not all(map(_has_content_and_role, data)):
+        messages = self.model.get("messages", [])
```
Comment (BenWilson2) [inline mlflow/openai/__init__.py:521]: What happens if a user wants to parameterize the prompts, but then wants to submit a list of string questions? Does the input of a mixed collection of `List[Union[Dict[str,str],str]]` work? Do we want  to support that use case?

Comment (harupy) [reply mlflow/openai/__init__.py:521]: I think schema enforcement rejectts it.

File: mlflow/openai/__init__.py (hunk 5 above)
Comment (harupy) [reply mlflow/openai/__init__.py:530]: got it, I'll update the message :)

File: mlflow/openai/__init__.py (hunk 6 above)
Comment (harupy) [reply mlflow/openai/__init__.py:482]: got it, I'll update the message :)

File: mlflow/openai/__init__.py (hunk 7 above)
Comment (harupy) [reply mlflow/openai/__init__.py:526]: > What happens if a user wants to parameterize the prompts, but then wants to submit a list of string questions?

Can you elaborate on this? If the prompts are parametrized, the parameters must be provided rather a list of string questions.

Comment (dbczumar) [reply mlflow/openai/__init__.py:526]: ^ This :)

File: mlflow/openai/__init__.py [hunk 8]
```diff
@@ -437,37 +504,44 @@ def __init__(self, model):
...
-                )
-            messages = data.to_dict(orient="records")
-        elif isinstance(data, list) and all(isinstance(d, dict) for d in data):
-            if not all(map(_has_content_and_role, data)):
-                raise mlflow.MlflowException.invalid_parameter_value(
-                    "Input list of dictionaries must contain keys 'content' and 'role'",
-                )
-            messages = data
+        messages = self.model.get("messages", [])
+        variables = _parse_variables(messages)
```
Comment (dbczumar) [inline mlflow/openai/__init__.py:527]: Do we have to do this parsing every time `predict()` is called, or can we parse once when the wrapper is constructed?

File: mlflow/openai/__init__.py [hunk 9]
```diff
@@ -191,6 +193,41 @@ def _log_secrets_yaml(local_model_dir, scope):
         yaml.safe_dump({e.value: f"{scope}:{e.secret_key}" for e in _OpenAIEnvVar}, f)
 
 
+def _parse_format_fields(s):
+    """
+    Parses format fields from a given string, e.g. "Hello {name}" -> ["name"].
+    """
+    return [fn for _, fn, _, _ in Formatter().parse(s) if fn is not None]
```
Comment (dbczumar) [inline mlflow/openai/__init__.py:200]: Amazing! Nice find!

**Review by dbczumar — APPROVED** [review]
LGTM with nits. Amazing work, @harupy ! Thanks so much!

File: mlflow/openai/__init__.py (hunk 8 above)
Comment (harupy) [reply mlflow/openai/__init__.py:527]: Great suggestion! 

> can we parse once when the wrapper is constructed?

yes, made this change :)

**Pull request merged**